### PR TITLE
AF-1654 and AF-1653: Adding C3 Charts

### PR DESCRIPTION
**Pull request by jesuino**
Add a new Renderer for Dashbuilder based on C3 charts, which should work offline.

### CODE REVIEW

**Comment by dgutierr** [comment]
@jesuino Great job William. Would you mind to record a short view to demo the new charts, attach it to the JIRA and share the link here in the PR?

**Comment by jesuino** [comment]
Hello @dgutierr - here's a screenshot of the charts on a page: https://issues.jboss.org/secure/attachment/12442494/c3_on_dashbuilder.png

Also, here's a screencast: https://issues.jboss.org/secure/attachment/12442493/c3_charts_implementation.ogv

Next tasks are:

1) implement filter
2) apply properties to charts

**Comment by jesuino** [comment]
@dgutierr It works very well on kie-wb - just need to apply filters and it should be possible to replace google with c3 https://issues.jboss.org/secure/attachment/12442597/jbom_wb_dashbuilderc3_demo.ogv

In any case, is there any reason for this PR for not be merged? This should not affect anything in the distributions (workbenches) because right now the workbenches will keep using Google (it is compiled with Google renderer, I just added a new option and I am not using it in jbpm-wb)

**Comment by dgutierr** [comment]
@jesuino @ederign Great job William!  

Personally, I'd rather wait after product branch is created before merging so as to make sure we do not impact current release. 

TL;DR; We can merge it next week. Waiting a few extra days won't hurt imho.

**Comment by jesuino** [comment]
@dgutierr @ederign Also included AF-1657 on this PR - Implemented the filters and before I refactored to use CDI just like the Google Renderer. Here's the filter implementation using the class FilterLabelSet: https://issues.jboss.org/secure/attachment/12442703/filters_c3.ogv

**Comment by Rikkola** [comment]
The version number went up so the GAV for dashbuilder-renderer-c3 module needs updating. Other than that this PR works fine after rebase.

**Comment by jesuino** [comment]
I had the version updated locally, @Rikkola 

There are two missing tasks which I will work on them soon: 

* Bind the chart properties to the C3 API: https://issues.jboss.org/browse/AF-1656
* Glue it into jbpm-wb and also have a system property that will allow users to keep using the Google Library if they want: https://issues.jboss.org/browse/AF-1655

**Comment by kiereleaseuser2** [comment]
Build finished. 5033 tests run, 11 skipped, 0 failed.

**Comment by kiereleaseuser2** [comment]
Build finished. 5033 tests run, 11 skipped, 0 failed.

**Comment by kiereleaseuser2** [comment]
Build finished. 5033 tests run, 11 skipped, 0 failed.

**Comment by kkufova** [comment]
@dgutierr, @ederign, could you please review the PR?

**Comment by jesuino** [comment]
@dgutierr  @ederign Properties were bound. This is what C3 looks like in Dashbuilder:

* Charts on a Page: https://issues.jboss.org/secure/attachment/12442494/c3_on_dashbuilder.png
* Adding C3 Charts: https://issues.jboss.org/secure/attachment/12442493/c3_charts_implementation.ogv
* C3 on Business Central: https://issues.jboss.org/secure/attachment/12442597/jbom_wb_dashbuilderc3_demo.ogv
* Filters on C3 Charts: https://issues.jboss.org/secure/attachment/12442703/filters_c3.ogv
* Properties: https://issues.jboss.org/secure/attachment/12444406/c3_binding_properties.ogv

After your review this could be merged so we can start working as making C3 default on Business Central and QE can start their work.

File: dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-c3/src/main/java/org/dashbuilder/renderer/c3/client/C3Renderer.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        return Arrays.asList(LINECHART, BARCHART, PIECHART, AREACHART, BUBBLECHART);
+    }
+    
+    @Override
+    public boolean isDefault(DisplayerType type) {
+//        return  BARCHART.equals(type) ||
+//                PIECHART.equals(type) ||
+//                AREACHART.equals(type) ||
+//                BUBBLECHART.equals(type) ||
+//                LINECHART.equals(type);
```
[the remark this answers is not in the document]
Comment (jesuino) [reply dashbuilder/dashbuilder-client/dashbuilder-renderers/dashbuilder-renderer-c3/src/main/java/org/dashbuilder/renderer/c3/client/C3Renderer.java]: I just removed it, later when implementing AF-1655[1] I will make it default again.

[1] https://issues.jboss.org/browse/AF-1655

**Comment by ederign** [comment]
Hi @kkufova could you please take a look at this? If you are not the QE assigned please let me know.

@jesuino this PR is to updating c3 on dashbuilder only right?

**Comment by jesuino** [comment]
@ederign right, it is only on dashbuilder, not on Business Central yet. It should be there when AF-1655 - I will work on AF-1655 once this PR is merged.

**Comment by kiereleaseuser2** [comment]
Build finished. 5033 tests run, 11 skipped, 0 failed.

**Comment by kiereleaseuser2** [comment]
Build finished. 5038 tests run, 11 skipped, 0 failed.

**Comment by kkufova** [comment]
@ederign, I would rather leave the review to an Automation Foundation QE. I already talked to the team and hopefully one of the QEs will take this.

**Comment by jhrcek** [comment]
War's won't help with this PR, as that functionality is not yet integrated in the workbench. I'm taking a look at this PR tomorrow (today I'm still on training).

**Comment by jesuino** [comment]
@jhrcek Thank you so much for your detailed analysis!

I was aware of some of these issues:

1) This is a C3 limitation, it does not support some legend position;
2) It is also how C3 works. It has no horizontal bars chart, so the solution is to rotate the bar chart. 
Both 1 and 2 were commented in https://issues.jboss.org/browse/AF-1656


Regarding: "in gwt charts toggling "X axis labels visible" only toggles the title of the axis, not the "values" on the axis"

That's a great observation, thanks. I will change C3 to also behave like that.

Regarding 3:

3.1 - Thanks, I will check what is happening;
3.2 That's another C3 limitation, it has no 3d Pie Chart;
3.3 That's a tough issue, I am still investigating, but I have a workaround in mind. It was captured in https://issues.jboss.org/browse/AF-1739

**Comment by jesuino** [comment]
@jhrcek I updated the PR with commits to address:

1) Stacked Bar charts not working;
2) X, Y axis values being hidden where only the text should be hidden.

As C3 limitation we have:

1: No 3d PieChart;
2) Not all legend positions are supported
3) The horizontal barchart is actually a rotated bar chart, which means that handling X axis will result in a vertical effect (for example: adding text to the axis) and Y axis will result in an horizontal effect.
4) C3 does not support all chart types, so maps are not supported at the moment, but this will change in future (see https://issues.jboss.org/browse/AF-1663)

As known issues we have:

1) Area charts are not displaying the area (see https://issues.jboss.org/browse/AF-1739)
2) The values for horizontal barchart seems messed - possibly find a way to make it look nicer.  

If I am missing something please let me know.

**Comment by kiereleaseuser2** [comment]
Build finished. 5046 tests run, 11 skipped, 0 failed.

**Comment by jesuino** [comment]
thanks @jhrcek I simply removed the TODOs because they were personal notes which I forgot to remove. 

@dgutierr @ederign Can we merge this? Please let me know if something else is needed on my side.

Next I am working on the Business Central integration with C3.

**Comment by kiereleaseuser2** [comment]
Build finished. 5050 tests run, 11 skipped, 0 failed.

**Comment by jhrcek** [comment]
@jesuino Great work! Thanks for the detailed summary of all the remaining tasks. I linked all of them to the original epic in Jira [AF-1504](https://issues.jboss.org/browse/AF-1504) so they're all accessible from a central place.
